File: src/leetcodejava/string/DetectCapital520.java
package leetcodejava.string;

import org.junit.Test;

/**
 * This is the solution of No. 520 problem in the LeetCode,
 * the website of the problem is as follow:
 * https://leetcode-cn.com/problems/detect-capital
 *
 * The description of problem is as follow:
 * ==========================================================================================================
 * 给定一个单词，你需要判断单词的大写使用是否正确。
 *
 * 我们定义，在以下情况时，单词的大写用法是正确的：
 *
 * 全部字母都是大写，比如"USA"。
 * 单词中所有字母都不是大写，比如"leetcode"。
 * 如果单词不只含有一个字母，只有首字母大写， 比如 "Google"。
 * 否则，我们定义这个单词没有正确使用大写字母。
 *
 * 示例 1:
 *
 * 输入: "USA"
 * 输出: True
 * 示例 2:
 *
 * 输入: "FlaG"
 * 输出: False
 * 注意: 输入是由大写和小写拉丁字母组成的非空单词。
 *
 * 来源：力扣（LeetCode）
 * 链接：https://leetcode-cn.com/problems/detect-capital
 * 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
 * ==========================================================================================================
 *
 * @author zhangyu (zhangyuyu417@gmail.com)
 */
public class DetectCapital520 {

    @Test
    public void testDetectCapital() {
        String word = "USA";
        boolean flag = detectCapitalUse(word);
        System.out.println(flag);
    }

    /**
     * 检测大小写字符串
     *
     * @param word 字符串
     * @return 布尔值
     */
    public boolean detectCapitalUse(String word) {
        if (word.length() <= 1) {
            return true;
        }
        boolean firstCharFlag = word.charAt(0) < 97;
        boolean secondCharFlag = word.charAt(1) < 97;

        if (!firstCharFlag && secondCharFlag) {
            return false;
        }
        for (int i = 2; i < word.length(); i++) {
            boolean charFlag = word.charAt(i) < 97;
            if (!secondCharFlag && charFlag) {
                return false;
            }
            if (secondCharFlag && !charFlag) {
                return false;
            }
        }
        return true;
    }
}

File: src/leetcodepython/string/detect_capital_520.py
# encoding='utf-8'

'''
/**
 * This is the solution of No.306 problem in the LeetCode,
 * the website of the problem is as follow:
 * https://leetcode-cn.com/problems/additive-number/
 * <p>
 * The description of problem is as follow:
 * ==========================================================================================================
 * 累加数是一个字符串，组成它的数字可以形成累加序列。
 * <p>
 * 一个有效的累加序列必须至少包含 3 个数。除了最开始的两个数以外，字符串中的其他数都等于它之前两个数相加的和。
 * <p>
 * 给定一个只包含数字 '0'-'9' 的字符串，编写一个算法来判断给定输入是否是累加数。
 * <p>
 * 说明: 累加序列里的数不会以 0 开头，所以不会出现 1, 2, 03 或者 1, 02, 3 的情况。
 * <p>
 * 示例 1:
 * <p>
 * 输入: "112358"
 * 输出: true
 * 解释: 累加序列为: 1, 1, 2, 3, 5, 8 。1 + 1 = 2, 1 + 2 = 3, 2 + 3 = 5, 3 + 5 = 8
 * 示例 2:
 * <p>
 * 输入: "199100199"
 * 输出: true
 * 解释: 累加序列为: 1, 99, 100, 199。1 + 99 = 100, 99 + 100 = 199
 * <p>
 * 来源：力扣（LeetCode）
 * 链接：https://leetcode-cn.com/problems/additive-number
 * 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
 * ==========================================================================================================
 *
 * @author zhangyu (zhangyuyu417@gmail.com)
 */
'''


class Solution:
    def detect_capital_use(self, word: str) -> bool:
        '''
            检查大小写
        Args:
            word: 单词
        Returns:
            布尔值
        '''
        if len(word) <= 1:
            return True
        first_char_flag = ord(word[0]) < 97
        second_char_flag = ord(word[1]) < 97
        if not first_char_flag and second_char_flag:
            return False
        for i in range(2, len(word)):
            char_flag = ord(word[i]) < 97
            if not second_char_flag and char_flag:
                return False
            if second_char_flag and not char_flag:
                return False
        return True

if __name__ == '__main__':
    s = 'USA'
    solution = Solution()
    result = solution.detect_capital_use(s)
    print(result)
    assert result == True
